File: array/40.py
from typing import List
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        if len(candidates) == 0 or target <= 0: return []
        # 回溯算法
        candidates.sort()
        ret = []
        self.lookBack(ret, candidates, target, [], 0)
        return ret

    def lookBack(self, ret, candidates, target, path, start_idx):
        if target == 0:
            ret.append(path)
            return
        i = start_idx
        while i < len(candidates):
            if i > start_idx and candidates[i - 1] == candidates[i]:
                i += 1
                continue
            candidate = candidates[i]
            if target - candidate >= 0:
                self.lookBack(ret, candidates, target-candidate, path + [candidate], i + 1)
            i += 1

def main():
    candidates = [10,1,2, 2,2,  7,6,5]
    target = 6
    print(Solution().combinationSum2(candidates, target))

if __name__ == "__main__":
    main()

File: leetcode-cpp/40.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <list>

using namespace std;

class Solution {
public:
    void lookBack(vector<int>& candidates, vector<vector<int>>& ret, list<int>& stk, int target, int idx){
//        for (auto i : stk){
//            cout << i << " ";
//        }
//        cout << "\t" << target<<endl;
        if (target == 0){
            vector<int> temp;
            temp.assign(stk.begin(), stk.end());
            ret.push_back(temp);
            return ;
        }
        for (int i = idx; i < candidates.size(); i++) {
            if (target < candidates[i])
                break;
            if (i == idx || (i != idx && candidates[i] != candidates[i - 1])) {
                stk.push_back(candidates[i]);
                lookBack(candidates, ret, stk, target - candidates[i], i+1);
                stk.pop_back();
            }
        }
    }

    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        if (candidates.size() == 0) return vector<vector<int>>();
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> ret;
        list<int> stk;
        lookBack(candidates, ret, stk, target, 0);
        return ret;
    }
};

int main(){
    Solution s;
    vector<int> candidates({2,5,2,1,2});
    int target = 5;
    vector<vector<int>> ret = s.combinationSum2(candidates, target);
    for (int i = 0; i < ret.size(); i++){
        for (int j = 0; j < ret[i].size(); j++){
            cout << ret[i][j] << "\t";
        }
        cout <<endl;
    }
}
